stepflow-rs/crates/stepflow-main/src/validate.rs

```diff
@@ -0,0 +1,288 @@
+    // Check for unused plugins (plugins not referenced by any routing rule)
+    for plugin_name in config.plugins.keys() {
+        let is_referenced = config.routing.rules.iter().any(|rule| rule.target == *plugin_name);
+        if !is_referenced {
+            println!(
+                "⚠️  Plugin '{}' is not referenced by any routing rule",
+                plugin_name
+            );
+            // This is a warning, not a failure
+        }
+    }
+
+    failures
+}
+
+/// Validate a workflow file
+async fn validate_flow(flow_path: &Path) -> Result<usize> {
+    println!("📋 Validating workflow: {}", flow_path.display());
+
+    // Load the workflow
+    let flow: Flow = load(flow_path)?;
+
+    // Run workflow validation using stepflow-analysis
+    let diagnostics = validate_workflow(&flow)
+        .change_context(MainError::ValidationError(
+            "Workflow validation failed".to_string(),
+        ))?;
+
+    // Count failures and display results
+    let (fatal_count, error_count, warning_count) = diagnostics.counts();
+    let failure_count = fatal_count + error_count;
+
+    // Display diagnostics
+    if diagnostics.is_empty() {
+        println!("✅ Workflow is valid");
+    } else {
+        println!(
+            "📊 Validation results: {} fatal, {} errors, {} warnings",
+            fatal_count, error_count, warning_count
+        );
+
+        // Group diagnostics by level
+        let fatal_diagnostics = diagnostics.at_level(DiagnosticLevel::Fatal);
+        let error_diagnostics = diagnostics.at_level(DiagnosticLevel::Error);
+        let warning_diagnostics = diagnostics.at_level(DiagnosticLevel::Warning);
+
+        // Print fatal diagnostics
+        if !fatal_diagnostics.is_empty() {
+            println!("\n🚨 Fatal Issues:");
+            for diagnostic in fatal_diagnostics {
+                print_diagnostic("FATAL", diagnostic);
+            }
+        }
+
+        // Print error diagnostics
+        if !error_diagnostics.is_empty() {
+            println!("\n❌ Errors:");
+            for diagnostic in error_diagnostics {
+                print_diagnostic("ERROR", diagnostic);
+            }
+        }
+
+        // Print warning diagnostics
+        if !warning_diagnostics.is_empty() {
+            println!("\n⚠️  Warnings:");
+            for diagnostic in warning_diagnostics {
+                print_diagnostic("WARN", diagnostic);
+            }
+        }
+    }
+
+    Ok(failure_count)
+}
+
+/// Print a formatted diagnostic message
+fn print_diagnostic(level: &str, diagnostic: &stepflow_analysis::Diagnostic) {
+    let path_str = if diagnostic.path.is_empty() {
+        String::new()
+    } else {
+        format!(" ({})", diagnostic.path.join("."))
+    };
+
+    println!("  {} {}{}", level, diagnostic.text, path_str);
+}
+
+/// Print a formatted error message
+fn print_error(context: &str, error: &error_stack::Report<MainError>) {
+    println!("❌ {}: {}", context, error);
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Could also add some tests to the `command` tests (these test the CLI and use snapshots on the output).